### ISSUE TITLE
ACL system improvment

### ISSUE THREAD

**Issue by Lothiraldan** [saltstack/salt#21679]
Following discussion on https://github.com/tinyclues/saltpad/issues/23, we
worked on the subject "How authentication system integrate with other User
Interfaces" during last Paris SaltStack sprint (
https://www.logilab.org/blogentry/288007).

We quickly saw that displaying user permissions in the User Interface (as
proposed in the saltpad ticket) works for simple ACL configurations like: 

```
external_auth:
  pam:
    thatch:
      - 'web*':
        - test.*
        - network.*
    steve:
      - .*
```

But will be more complex with ACL configurations using grains, pillar, cidr or 
range cluster. The User Experience with these configurations will be less than
ideal and quite painful. 

So, we tried to see how the ACL system could be enhanced to be able to
target for example all minions (`*`) and get the job running only on
minions we have rights on.

We first tried to modify the Web UI (we tried on saltpad) to dynamically
change the target to include the permissions list and let's the targeting
system do the trick. For example, with this ACL configuration:

```
external_auth:
  pam:
    foo:
      - 'devel*':
        - test.*
        - network.*
      - 'preprod*'
        - test.ping
```

When launching a job on `*`, the Web UI changes the matcher from glob to
compound and sends this target instead: `* and ( devel* or preprod* )`. 
Technically it worked, we only get results from minions we have rights on.

But here are the downsides:
- it requires some work on each client and may require upgrade if Salt adds new
  matchers.
- as we're modifying the targets of jobs, the new targets are displayed, so 
  reporting will be more difficult as the human reader need to isolate the real
  target part from the permission part
- changing the targets is ugly and error prone 

![capture d ecran 2015-03-08 a 19 55 18](https://cloud.githubusercontent.com/assets/243665/6658415/ed496fd0-cb68-11e4-8469-bb2e93221468.png)

Funny side-effect too, in SaltPad we have a "Redo job" button, and with the
modification it modifies the target each time we redo a job, so targets keep 
growing and growing.

![capture d ecran 2015-03-08 a 20 27 06](https://cloud.githubusercontent.com/assets/243665/6658418/f2b2cbd8-cb68-11e4-91ed-2cbb30dd4906.png)

Here are the changes for reference:
https://github.com/tinyclues/saltpad/compare/compound-perms

Then we dug into the ACL implementation and saw it was implemented on
the master side. It's a pretty big piece of code that does a lot of
work. I don't have a big cluster to test ACL system on, but the current
ACL implementation seems likely to be a problem for scaling. 

Moreover the current implementation on the master side make some matchers
impossible to use with ACL (
https://github.com/saltstack/salt/blob/develop/salt/utils/minions.py#L545-L557).
If we have this configuration:

```
external_auth:
  pam:
    foo:
      - 'S@10.0.2.0/24':
        - *.*
```

We will only be able to launch job using this target `-S '10.0.2.0/24'` and
not `-S '10.0.2.15/32'` making the cidr matcher support in ACL much less
useful.

It would be possible to modify the master to do exactly the same trick
we implemented in SaltPad, but we will have exactly the same issues 
except that the clients will not need to implement it and it will be
always up-to-date with matchers.

The ACL implementation will anyway reside on the master side for runner and
wheel ACL.

Then we looked to see if the ACL checks could be moved on the minion side. 
The code checking if the minion should execute or not the job is pretty
concise and can already do a compound match. It would be very easy to send
the user permission (master gets it here 
https://github.com/saltstack/salt/blob/develop/salt/master.py#L2184-L2186)
with the job load and ask the minion to check if one permission with the
selected function match or else return, exactly like when the targeting
doesn't match the current minion.

We saw that the master does some pre-processing before sending jobs to 
minions, for list matcher for example (the minion claims that the master do
some pre-processing for pillar based targeting, couldn't find the code
which does that), but it should not be a problem as ACL permissions will
only filter out some minions.

But the sky is not so shiny, this solution has a big problem and a little
one. Currently, the master pre-filter minions before sending the job to
send an event with the list of minions (here it gets the minion list
https://github.com/saltstack/salt/blob/develop/salt/master.py#L2336-L2341
and send it here
https://github.com/saltstack/salt/blob/develop/salt/master.py#L2366-L2379).
I'm not sure how it should work and if currently works with CIDR matcher,
but it will likely to be either removed or modified in order to not
announce jobs with minions which will be filtered out later.

The biggest problem is the backward compatibility. If we implemented this
solution and in the next version asks the minions to do the ACL part, it
will fail with outdated minions. One solution could be to enable this
solution only if all minions are up-to-date. Or by adding a grain in new
salt minions version and check before if all minions have this grain on or
not. Couldn't find a good solution for this problem.

The proposed solution is to try to refactor the ACL system to move the
biggest part on the minion side after finding a correct way to handle
backward compatibility. It's not a final proposition, it's based on our
last thought about the ACL system design and all contributions and
thoughts will be more than welcome on the subject.

**Comment by jfindlay** [comment]
@Lothiraldan, thanks for working through this.  Any code updates are always welcome.

**Comment by Lothiraldan** [comment]
@jfindlay Before starting to code, the PR will not be that huge I think, I would prefer to have opinions of @whiteinge and maybe @thatch45 as it's quite a big design change and how we could handle backward compatibility is not yet decided.

**Comment by thatch45** [comment]
Ping me if I don't get back to this soon. @cachedout is also working on re-factoring this code

**Comment by Lothiraldan** [comment]
Great I would be glad to help, I want only to agree on the new design before starting to code.

**Comment by whiteinge** [comment]
I'd be interested in @cachedout 's ideas on this while still maintaining backward-compat.

Could a new `expr_form` be added that is eauth-specific? That would allow the existing eauth targeting semantics to stay in-tact but allow for an alternate targeting that parses the eauth config to provide the informed guess of the subset of authorized minions.

Perhaps `salt -a pam -A '*' test.ping` and `local_client.cmd('*', 'test.ping', expr_form='eauth')`.

Are there any other eauth-specific kinds of targeting to take into account here besides `*` ?

**Comment by Lothiraldan** [comment]
@whiteinge In my idea, every kind of targeting should works with the minion-side eauth. It should be possible to target `-S '10.0.2.0/24'` and run the job only on authorized minions.

**Comment by Lothiraldan** [comment]
Ping @basepi from pycon!

**Comment by basepi** [comment]
@Lothiraldan I kind of feel like what we talked about at PyCon (specifically finding a way to secure certain grains and whatnot for pillar targeting) might be worth opening a new issue for. *shrugs* But thanks for the ping.

**Comment by Lothiraldan** [comment]
Yes definitely, I will open a separate issue, the ping was for getting @thatch45 or @cachedout on this issue

**Comment by Lothiraldan** [comment]
@thatch45 @cachedout Do you want me to start a pull-request with the idea proposed?

**Comment by Lothiraldan** [comment]
@thatch45 @cachedout Any progression about this feature?

**Comment by thatch45** [comment]
You have already found the problem, we restrict execution on the master but resolve execution on the minion.
What I think might make sense is to do it minion side, allow for the minion to have a perms resolver. We have quite a few people who want minion side perm resolution. The way to do this would be to send the configuration down to the minion used by the auth and then restrict auth on the master by only checking the yes/no from eauth/client_acl.

You also pointed out the main issue with this being that the master needs to pre-allocate a list of minions it expects to return. So I think that the minions that reject the command would need to send a job return to the master stating that they received the job request but that they rejected it. This would also give us the full audit trail knowing exactly where the job was rejected
